pkg/controller/deployment/deployment_controller.go

```diff
@@ -355,6 +371,44 @@ func (dc *DeploymentController) syncDeployment(key string) error {
 		return err
 	}
 
+	if dc.garbageCollectorEnabled {
+		rsList, err := dc.rsLister.ReplicaSets(deployment.Namespace).List(labels.Everything())
+		if err != nil {
+			glog.Errorf("Error getting ReplicaSets for deployment %q: %v", deployment, err)
+			return err
+		}
+
+		deploymentSelector, err := unversioned.LabelSelectorAsSelector(deployment.Spec.Selector)
+		if err != nil {
+			return fmt.Errorf("deployment %s/%s has invalid label selector: %v", deployment.Namespace, deployment.Name, err)
+		}
+		cm := controller.NewReplicaSetControllerRefManager(dc.rsControl, deployment.ObjectMeta, deploymentSelector, getDeploymentControllerKind())
+		matchesAndControlled, matchesNeedsController, controlledDoesNotMatch := cm.Classify(rsList)
+		for _, replicaSet := range matchesNeedsController {
+			err := cm.AdoptReplicaSet(replicaSet)
```

Comment:
These replica sets are coming from a shared cache so you need to deep-copy before mutating them.